lib/Controller/WhiteboardController.php

```diff
@@ -106,4 +85,52 @@ public function show(int $fileId): DataResponse {
 			'data' => $data,
 		]);
 	}
+
+	private function verifySharedToken(string $token, int $fileId): bool {
+		[$roomId, $timestamp, $signature] = explode(':', $token);
+
+		if ($roomId !== (string)$fileId) {
+			return false;
+		}
+
+		$sharedSecret = $this->configService->getWhiteboardSharedSecret();
+		$payload = "{$roomId}:{$timestamp}";
+		$expectedSignature = hash_hmac('sha256', $payload, $sharedSecret);
+
+		return hash_equals($expectedSignature, $signature);
+	}
+
+	/**
+	 * @throws NotPermittedException
+	 * @throws NoUserException
+	 * @throws JsonException
+	 */
+	#[NoAdminRequired]
+	#[NoCSRFRequired]
+	#[PublicPage]
+	public function update(int $fileId, array $data): DataResponse {
+		$whiteboardAuth = $this->request->getHeader('X-Whiteboard-Auth');
+		$whiteboardUser = $this->request->getHeader('X-Whiteboard-User');
+
+		if (!$whiteboardAuth || !$this->verifySharedToken($whiteboardAuth, $fileId)) {
+			return new DataResponse(['message' => 'Unauthorized'], Http::STATUS_UNAUTHORIZED);
+		}
+
+		$user = $this->userManager->get($whiteboardUser);
+
+		if (!$user) {
+			return new DataResponse(['message' => 'Invalid user'], Http::STATUS_BAD_REQUEST);
+		}
+
+		$userFolder = $this->rootFolder->getUserFolder($user->getUID());
+		$file = $userFolder->getById($fileId)[0];
```

Comment:
There are a few things we should check here when getting the file, best see 

https://github.com/nextcloud/text/blob/74aa530fb4ca922d3577b64cc01676d28c449988/lib/Service/DocumentService.php#L496-L524

for all edge cases that may occur 🙈